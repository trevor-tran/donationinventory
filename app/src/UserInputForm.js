import { useFormik } from 'formik';
import dayjs from 'dayjs';
import * as yup from 'yup';
import {

  TextField,
  Button,
  MenuItem,
  InputLabel,
  Select,
  FormControl,
  InputAdornment,
} from '@mui/material';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';
import axios from 'axios';

const validationSchema = yup.object({
  donor: yup
    .string("Enter donor's name")
    .matches(/^[a-zA-Z]+$/, "Please enter a valid name")
    .required("Donor's name is required"),
  amount: yup
    .number("Enter donation amount")
    .min(1)
    .required('Donation amount is required'),
});

/**
 * Add/Update Donation form.
 * Inputs are validated where nesscessary
 */
export default function UserInputForm(props) {
  const URL = "http://localhost:8080/donation";

  const formik = useFormik({
    initialValues: {
      id: props.values?.id || null,
      donor: props.values?.donor || '',
      type: props.values?.type || 'Money',
      amount: props.values?.amount || '',
      date: props.values?.date ? dayjs(props.values?.date) : dayjs()
    },
    validationSchema: validationSchema,
    onSubmit: async (values) => {
      const data = { ...values, date: dayjs(values.date).format("YYYY-MM-DD") };
      const response = await axios.post(URL, data);
      props.onSave(response.data);
    }
  });
  return (
    <form className={"d-flex justify-content-center align-items-center" + (props.dialog ? " flex-column mr-auto" : "")} onSubmit={formik.handleSubmit}>
      <TextField
        fullWidth
        className="m-3"
        id="donor"
        name="donor"
        label="Donor"
        value={formik.values.donor}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        error={formik.touched.donor && Boolean(formik.errors.donor)}
        helperText={formik.touched.donor && formik.errors.donor}
      />
      <FormControl className="m-3" fullWidth>
        <InputLabel id="type-of-donation-label">Type of Donation</InputLabel>
        <Select
          labelId="type-of-donation-label"
          id="type-of-donation-select"
          value={formik.values.type}
          name="type"
          label="Type of Donation"
          onChange={formik.handleChange}
        >
          <MenuItem value="Money">Money</MenuItem>
          <MenuItem value="Food">Food</MenuItem>
          <MenuItem value="Clothing">Clothing</MenuItem>
        </Select>
      </FormControl>
      <TextField
        fullWidth
        className="m-3"
        id="amount"
        name="amount"
        label="Amount"
        type="number"
        min={1}
        value={formik.values.amount}
        onChange={formik.handleChange}
        onBlur={formik.handleBlur}
        InputProps={{
          startAdornment: <InputAdornment position="start">{formik.values.type === "Money" ? "$" : "lb"}</InputAdornment>,
          inputProps: { min: 0
        }
        }}
        error={formik.touched.amount && Boolean(formik.errors.amount)}
        helperText={formik.touched.amount && formik.errors.amount}
      />
      <div className="m-3 w-100">
      <LocalizationProvider dateAdapter={AdapterDayjs}>
        <DatePicker label="Date of Donation"
          name="date"
          value={formik.values.date}
          onChange={date => formik.setFieldValue("date", date, false)}
          maxDate={dayjs()}
        />
      </LocalizationProvider>
      </div>
      <div style={{textAlign: "center"}}>
      <Button color="primary" variant="contained" type="submit"> Save </Button>
      </div>
    </form>
  )
}
