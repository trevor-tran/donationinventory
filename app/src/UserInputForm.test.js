import renderer from 'react-test-renderer';
import UserInputForm from './UserInputForm';

describe("UserInputForm Page", () => {

  it("verifies against its snapshot", () => {
    const tree = renderer
      .create(<UserInputForm />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })
})