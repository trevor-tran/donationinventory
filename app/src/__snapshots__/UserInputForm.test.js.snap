// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInputForm Page verifies against its snapshot 1`] = `
<form
  className="d-flex justify-content-center align-items-center"
  onSubmit={[Function]}
>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root m-3 css-wb57ya-MuiFormControl-root-MuiTextField-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={false}
      htmlFor="donor"
      id="donor-label"
    >
      Donor
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="donor"
        name="donor"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-yjsfm1"
        >
          <span>
            Donor
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth m-3 css-q8hpuo-MuiFormControl-root"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      id="type-of-donation-label"
    >
      Type of Donation
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
      onClick={[Function]}
    >
      <div
        aria-controls=":r1:"
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="type-of-donation-label type-of-donation-select"
        className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="type-of-donation-select"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
        tabIndex={0}
      >
        Money
      </div>
      <input
        aria-hidden={true}
        aria-invalid={false}
        className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        disabled={false}
        name="type"
        onAnimationStart={[Function]}
        onChange={[Function]}
        required={false}
        tabIndex={-1}
        value="Money"
      />
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Type of Donation
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root m-3 css-wb57ya-MuiFormControl-root-MuiTextField-root"
    min={1}
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink={true}
      htmlFor="amount"
      id="amount-label"
    >
      Amount
    </label>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart css-8j6b76-MuiInputBase-root-MuiOutlinedInput-root"
      onClick={[Function]}
    >
      <div
        className="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
        >
          $
        </p>
      </div>
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
        disabled={false}
        id="amount"
        min={0}
        name="amount"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="number"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-14lo706"
        >
          <span>
            Amount
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="m-3 w-100"
  >
    <div
      className="MuiFormControl-root MuiTextField-root css-z3c6am-MuiFormControl-root-MuiTextField-root"
      onClick={[Function]}
      onMouseUp={[Function]}
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink={true}
        htmlFor=":r3:"
        id=":r3:-label"
      >
        Date of Donation
      </label>
      <div
        className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          autoComplete="off"
          autoFocus={false}
          className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
          disabled={false}
          id=":r3:"
          inputMode="text"
          name="date"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPaste={[Function]}
          placeholder="MM/DD/YYYY"
          readOnly={false}
          required={false}
          type="text"
          value="03/26/2024"
        />
        <div
          className="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            aria-label="Choose date, selected date is Mar 26, 2024"
            className="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CalendarIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
              />
            </svg>
          </button>
        </div>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-14lo706"
          >
            <span>
              Date of Donation
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="submit"
    >
       Save 
    </button>
  </div>
</form>
`;
